src/cairoUtilFuncGen/abi/abiDecode.ts

```diff
@@ -390,44 +431,69 @@ export class AbiDecode extends StringIndexedFuncGenWithAuxiliar {
       `}`,
     ].join('\n');
 
-    this.requireImport('starkware.cairo.common.uint256', 'Uint256');
-    this.requireImport('warplib.memory', 'wm_index_dyn');
-    this.requireImport('warplib.maths.utils', 'felt_to_uint256');
-    this.requireImport('warplib.maths.utils', 'narrow_safe');
+    const importedFuncs = [
+      this.requireImport('starkware.cairo.common.uint256', 'Uint256'),
+      this.requireImport('warplib.memory', 'wm_index_dyn'),
+      this.requireImport('warplib.maths.utils', 'felt_to_uint256'),
+      this.requireImport('warplib.maths.utils', 'narrow_safe'),
+    ];
+
+    const funcInfo = {
+      name,
+      code,
+      functionsCalled: [...importedFuncs, ...functionsCalled, writeToMemFunc],
+    };
+    const generatedFunc = this.createAuxiliarGeneratedFunction(funcInfo);
 
-    this.auxiliarGeneratedFunctions.set(key, { name, code });
-    return name;
+    this.auxiliarGeneratedFunctions.set(key, generatedFunc);
+    return generatedFunc;
   }
 
-  private createStructDecoding(type: UserDefinedType, definition: StructDefinition) {
+  private createStructDecoding(
+    type: UserDefinedType,
+    definition: StructDefinition,
+  ): CairoFunctionDefinition {
     const key = type.pp();
     const existing = this.auxiliarGeneratedFunctions.get(key);
-    if (existing !== undefined) return existing.name;
+    if (existing !== undefined) return existing;
 
     let indexWalked = 0;
     let structWriteLocation = 0;
-    const instructions = definition.vMembers.map((member, index) => {
-      const type = generalizeType(safeGetNodeType(member, this.ast.inference))[0];
-      const elemWidth = CairoType.fromSol(type, this.ast, TypeConversionContext.Ref).width;
-      const decodingCode = this.generateDecodingCode(
-        type,
-        'mem_index',
-        `member${index}`,
-        `${indexWalked}`,
-      );
-      indexWalked += Number(getByteSize(type, this.ast.inference));
-      structWriteLocation += index * elemWidth;
-      const getMemLocCode = `let mem_to_write_loc = ${add('struct_ptr', structWriteLocation)};`;
-      const writeMemLocCode = `${this.memoryWrite.getOrCreate(
-        type,
-      )}(mem_to_write_loc, member${index});`;
-      return [
-        `// Decoding member ${member.name}`,
-        `${decodingCode}`,
-        `${getMemLocCode}`,
-        `${writeMemLocCode}`,
-      ].join('\n');
-    });
+    const decodingInfo: [string, CairoFunctionDefinition[]][] = definition.vMembers.map(
+      (member, index) => {
+        const type = generalizeType(safeGetNodeType(member, this.ast.inference))[0];
```

Comment:
Let's use destructuring assignment instead of accessing by index. `const [type] = `